### PR TITLE
Add new recognition observers: on_end() and on_post_recognition()

### PR DESCRIPTION
**Pull request by daanzu**
@Danesprite Needs work on the natlink & sphinx backends.

And probably general look over and testing.

### CODE REVIEW

**Comment by Danesprite** [comment]
Thanks for this. I'll get it working with Natlink and Sphinx. Do you think extra arguments should be added to `engine.do_recognition()` for these new callbacks?

**Comment by daanzu** [comment]
Looks good.

Yeah, I'm not sure about the rule argument. Actually, ideally, I would want it to be optional (only passed if the receiver has an argument named `rule`) passed to the `on_recognition` callback. I'm not sure how much work it would be to implement that currently.

**Comment by Danesprite** [comment]
It's certainly possible to make `rule` an optional argument. It seems to me that any use of the `rule` argument would be better handled in a custom rule/grammar class instead.

**Comment by daanzu** [comment]
My thinking about the `rule` argument is for a GUI I'm working on, so I can easily display the grammar and rule of the recognition for debugging.

**Comment by Danesprite** [comment]
Okay then, that's a good use case. To expand on this, perhaps the argument should be the `extras` dictionary used internally by `rule.process_recognition()`?That would contain the corresponding grammar and rule objects, plus the parse tree node used to match/decode words to a specific rule and mapping as well as any other keys added by the rule implementations:
```Python
extras = {
    "_grammar":  self.grammar,
    "_rule":     self,
    "_node":     node,
}
```

The `process_recognition()` method for `MappingRule` and `CompoundRule` could be changed to return these internal `extra` dictionaries after calling `self._process_recognition()`.

**Comment by daanzu** [comment]
Oh, I forgot about the `extras`. That would be great.

**Comment by LexiconCode** [comment]
Would it be possible for me to add support this with casters  rules as well?

**Comment by daanzu** [comment]
Actually, I think I see a cleaner way to implement this. I'll give it a try tomorrow night.

**Comment by daanzu** [comment]
How does that look?

File: dragonfly/engines/base/recobs.py [hunk 1]
```diff
@@ -77,15 +78,20 @@ def notify_begin(self):
                                     " method of recognition observer %s: %s"
                                     % (observer, e))
 
-    def notify_recognition(self, words):
+    def notify_recognition(self, words, rule, node):
         for observer in self._observers:
             try:
                 if hasattr(observer, "on_recognition"):
-                    observer.on_recognition(words)
+                    (arg_names, _, kwargs_name, _) = inspect.getargspec(observer.on_recognition)
```
Comment (Danesprite) [inline dragonfly/engines/base/recobs.py]: Lines like this are causing a lot of deprecation warnings when running the tests in Python 3. I think we should use `inspect.getargspec()` for Python 2.7 and `inspect.getfullargspec()` in Python 3.

Comment (Danesprite) [reply dragonfly/engines/base/recobs.py]: Fixed. There's another `getargspec()` deprecation warning in `action_function.py` that I'll fix separately.

**Comment by Danesprite** [comment]
I'm glad you thought of just passing the `Node` object instead of `extras`. It is much cleaner and it wouldn't be difficult to use it to generate internal `MappingRule` `extras` dictionaries. Do you think this is ready to merge now?

**Comment by Danesprite** [comment]
BTW @LexiconCode you can use the `Node` object to get at the matching spec(s). For a non-CCR `MappingRule`, you can get the matching `Compound` element by doing this:

```Python
def on_post_recognition(words, node):
    element = node.children[0].children[0].actor
    print(element._spec)

```

**Comment by daanzu** [comment]
Yeah, once thought about it, this seemed nicer. Thanks for the fix ups. Should be good to merge as far as I can see!

**Pull request merged**